### PR TITLE
Nilability checking improvements #2

### PR DESCRIPTION
**Pull request by vasslitvinov**
* Implement some previously-missing checks.
* Improve the error messages in some cases.
* Minor fixes in two existing tests.
* Add testing of all combinations of nilability and memory management of lhs and rhs when initializing a variable this way:
```chpl
        var lhs: <lhs type> = rhs;
```
The tests are added under:

    test/classes/errors/nilability-init-var/

Some improvements are implemented as additional `init=` methods or
otherwise in the modules. Other improvements are from compiler changes.

In some cases, the following code:
```chpl
    var lhs: <lhs type> = rhs;
    compilerError("something");
```
results in compilerError() firing, whereas without that call
the compiler would report an error on the `var lhs` line.
This is because the `var lhs` error would be reported in
`insertAndResolveCasts(fn)`, which is called after
`resolveBlockStmt(fn->body)` completes successfully.

Some of the compiler changes in this PR make the compiler still report
the `var lhs` error even in the presence of a subsequent compilerError().

The all-combinations tests were created by running this script:

    classes/errors/nilability-generate.chpl

then manually adjusting two files to comment out compilerError(), see below.
I set to test them with --no-codegen, to save testing time. This is a
consideration because 28 of them compile without errors.

FUTURE WORK

* Remove the error upon `var lhs: <shared?> = <owned!>`, which is legal.
  See `init-var-oknil-shared-from-nonnil-owned.future`

* The following cases would still fire compilerError() despite an error
  in the preceding `var lhs` line. I commented out compilerError() in those
  tests to trigger the desired error messages. We want to see those messages
  ahead of compilerError().

    init-var-nonnil-borrowed-from-oknil-nil.chpl
    init-var-nonnil-unmanaged-from-oknil-nil.chpl

* In #11118 we preferred the error message pattern `cannot assign to <lhs>  from <rhs>`.
  There are a few places in the modules and in the compiler that
  either switch the order or say `assign <lhs> to <rhs>`.
  It would be nice to make those consistent with #11118.

Testing:
* [x] linux64 -futures
* [x] gasnet multilocale tests
* [x] numa
* [x] retest after rebasing from #13641

### CODE REVIEW

File: modules/internal/OwnedObject.chpl [hunk 1]
```diff
@@ -245,16 +245,37 @@ module OwnedObject {
        refer to `nil` after this call.
      */
     proc init=(pragma "leaves arg nil" pragma "nil from arg" ref src:_owned) {
+      if isNonNilableClass(this.type) && isNilableClass(src) &&
+         !chpl_legacyNilClasses
+      then
+        compilerError("cannot assign to a non-nilable owned variable from a nilable class");
```
Comment (mppf) [inline modules/internal/OwnedObject.chpl]: Should this error say "cannot copy-initialize a ... from a ..." here and elsewhere?

File: test/classes/errors/nilability-init-var/init-var-nonnil-owned-from-nonnil-borrowed.good [hunk 2]
```diff
@@ -0,0 +1 @@
+init-var-nonnil-owned-from-nonnil-borrowed.chpl:9: error: cannot assign to an owned variable from a borrowed class
```
Comment (mppf) [inline test/classes/errors/nilability-init-var/init-var-nonnil-owned-from-nonnil-borrowed.good]: This error message makes it sound like the offending code had `lhs = borrowed MyClass`

File: modules/internal/OwnedObject.chpl [hunk 3]
```diff
@@ -245,16 +245,37 @@ module OwnedObject {
...
+
       // Use 'this.type.chpl_t' in case RHS is a subtype
       this.chpl_t = this.type.chpl_t;
       this.chpl_p = src.release();
+      this.complete();
+    }
+
+    proc init=(src: shared) {
+      compilerError("cannot assign to an owned variable from a shared class");
+      this.chpl_t = int; //dummy
```
Comment (lydia-duncan) [inline modules/internal/OwnedObject.chpl]: If we're just going to compilerError, are these lines necessary?

File: modules/internal/OwnedObject.chpl [hunk 4]
```diff
@@ -245,16 +245,37 @@ module OwnedObject {
...
+    }
+
+    proc init=(src: borrowed) {
+      compilerError("cannot assign to an owned variable from a borrowed class");
+      this.chpl_t = int; //dummy
+    }
+
+    proc init=(src: unmanaged) {
+      compilerError("cannot assign to an owned variable from an unmanaged class");
```
Comment (lydia-duncan) [inline modules/internal/OwnedObject.chpl]: In this and the above errors, I would maybe end with "class *instance*"

File: modules/internal/OwnedObject.chpl (hunk 1 above)
Comment (lydia-duncan) [reply modules/internal/OwnedObject.chpl]: If that felt too verbose, could also say "cannot *create* a ... from a ..."

File: modules/internal/OwnedObject.chpl (hunk 3 above)
Comment (lydia-duncan) [reply modules/internal/OwnedObject.chpl]: And if the compiler is yelling about them, that seems like something we should create an issue for

Comment (vasslitvinov) [reply modules/internal/OwnedObject.chpl]: Created #13688.

**Pull request merged**